src/main/java/com/android/volley/Request.java

```diff
@@ -83,8 +83,12 @@
     /** Default tag for {@link TrafficStats}. */
     private final int mDefaultTrafficStatsTag;
 
+    /** Lock to guard state which can be mutated after a request is added to the queue. */
+    private final Object mLock = new Object();
```

Comment:
This would be a pretty rough API for Request subclasses to implement, and it would mean losing control of the lock so subclasses could end up doing all kinds of bad things with the lock held, leading to the concerns you mentioned earlier about open-ended locks.

Most importantly, though, I don't think it strengthens any guarantees as noted in my other comment.

Reply:
I don't think it is no more difficult to implement than what your subclasses have already implemented, and in some ways is simpler -- in that they don't have to create their own lock.

Designing for extension is difficult, and even more difficult when locks are involved. But I don't think you can mitigate by hiding the cancellation lock from subclasses -- that just prevents them from correctly maintaining the state.